### PR TITLE
fix: Fix a rounding error at the `Number::maxRep` cusp

### PR DESCRIPTION
**Pull request by ximinez**
## High Level Overview of Change

If the mantissa result of an operation is `maxRep`, and should round up, it rounds down instead.

### CODE REVIEW

**Comment by godexsoft** [comment]
We recently merged a refactor to `develop` that enables clang-tidy's `readability-identifier-naming`. Your branch now has heavy conflicts that are largely mechanical. Below is a workflow that aligns your branch's naming with develop _before_ merging, which should minimize the merge conflicts.

## One-time setup

If you don't already have clang-tidy working in your env, on macOS:

```bash
brew install llvm@21
# Follow brew's hint to put $(brew --prefix llvm@21)/bin on PATH so run-clang-tidy is found.
```

## Workflow on your branch (before merging develop)

**1. Grab the new `.clang-tidy` from develop without pulling anything else.** Sync your fork on GitHub first, then:

```bash
git remote -v   # should show 'upstream' among others; if not:
# git remote set-url upstream git@github.com:XRPLF/rippled.git
git fetch upstream
git checkout upstream/develop -- .clang-tidy
```

**2. Reconfigure conan/cmake so `compile_commands.json` is fresh.**

**3. Apply renames for the files modified in your PR:**

```bash
git diff --name-only $(git merge-base HEAD upstream/develop) HEAD \
  | grep -E '\.(cpp|h|hpp|ipp)$' \
  | xargs run-clang-tidy -p build -fix -allow-no-checks
# or -p .build, or whatever your build dir is called
```

**4. Build + test, then commit as a single dedicated commit:**

```bash
cmake --build build -j8
git commit -am "refactor: Align identifier naming with develop"
```

**5. Now merge develop:**

```bash
git merge upstream/develop
```

## Extra

Run clang-tidy once more after the merge to catch any stragglers introduced from develop's side:

```bash
run-clang-tidy -p build -fix -allow-no-checks src tests
# or -p .build, or whatever your build dir is called
```

File: src/test/app/AMMExtended_test.cpp [hunk 1]
```diff
@@ -65,11 +65,16 @@ namespace xrpl::test {
...
+class AMMExtended_test : public jtx::AMMTest
 {
     // Use small Number mantissas for the life of this test.
     NumberMantissaScaleGuard const sg{xrpl::MantissaRange::MantissaScale::Small};
 
+    // For now, just disable SAV entirely, which locks in the small Number
+    // mantissas
+    FeatureBitset const all_{
+        testableAmendments() - featureSingleAssetVault - featureLendingProtocol - fixCleanup3_2_0};
+
```
[the remark this answers is not in the document]
Comment (ximinez) [reply src/test/app/AMMExtended_test.cpp:77]: #no

File: src/libxrpl/protocol/Rules.cpp [hunk 2]
```diff
@@ -38,11 +38,21 @@ setCurrentTransactionRules(std::optional<Rules> r)
     // Make global changes associated with the rules before the value is moved.
     // Push the appropriate setting, instead of having the class pull every time
     // the value is needed. That could get expensive fast.
-    bool const enableLargeNumbers =
-        !r || (r->enabled(featureSingleAssetVault) || r->enabled(featureLendingProtocol));
+
+    // The improved accuracy that will come from enabling large
+    // mantissas is a goal separate from SAV and LP. It was originally
+    // only tied to those two amendments to avoid needing a new
+    // amendment of it's own, and because they require that behavior.
```
[the remark this answers is not in the document]
Comment (ximinez) [reply src/libxrpl/protocol/Rules.cpp]: Yeah, that's a typo.

File: src/test/basics/Number_test.cpp [hunk 3]
```diff
@@ -6,6 +6,8 @@
 #include <xrpl/protocol/SystemParameters.h>
 #include <xrpl/protocol/XRPAmount.h>
```
[the remark this answers is not in the document]
Comment (ximinez) [reply src/test/basics/Number_test.cpp:8]: > which typically does not provide the cpp_int typedef. 

Apparently it does, because the CI builds are (mostly) working. clang-tidy also forced removing `cpp_int`. Robot fight!

File: src/libxrpl/protocol/Rules.cpp [hunk 4]
```diff
@@ -38,11 +38,21 @@ setCurrentTransactionRules(std::optional<Rules> r)
...
+    // mantissas even in the absence of the other two amendments.
+    bool const enableCuspRoundingFix = !r || r->enabled(fixCleanup3_2_0);
+    bool const enableVaultNumbers = enableCuspRoundingFix ||
+        (r->enabled(featureSingleAssetVault) || r->enabled(featureLendingProtocol));
     Number::setMantissaScale(
-        enableLargeNumbers ? MantissaRange::MantissaScale::Large
-                           : MantissaRange::MantissaScale::Small);
+        enableCuspRoundingFix ? MantissaRange::MantissaScale::Large
+                              : (enableVaultNumbers ? MantissaRange::MantissaScale::LargeLegacy
+                                                    : MantissaRange::MantissaScale::Small));
```
[the remark this answers is not in the document]
Comment (ximinez) [reply src/libxrpl/protocol/Rules.cpp]: Good catch!

[the remark this answers is not in the document]
Comment (ximinez) [reply src/libxrpl/protocol/Rules.cpp]: > Consider updating the transaction-phase guard logic to also apply Rules (or at least the mantissa scale selection) when `fixCleanup3_2_0` is enabled without SAV/LP.

Fixed in e22938d69f

File: include/xrpl/basics/Number.h [hunk 5]
```diff
@@ -90,15 +117,35 @@ struct MantissaRange
...
+        {
+            case MantissaScale::Small:
+            case MantissaScale::LargeLegacy:
+                return CuspRoundingFix::Disabled;
+            case MantissaScale::Large:
+                return CuspRoundingFix::Enabled;
+            default:
+                // If called in a constexpr context, this throw assures that the build fails if an
                 // invalid scale is used.
                 throw std::runtime_error("Unknown mantissa scale");
```
Comment (Tapanito) [inline include/xrpl/basics/Number.h]: `codecov` is complaining that the `default` case is not tested. It seems to me testing this path would be non-trivial. Is it worth adding `//LCOV_EXCL_LINE?


```suggestion
                throw std::runtime_error("Unknown mantissa scale"); //LCOV_EXCL_LINE
```

File: include/xrpl/basics/Number.h [hunk 6]
```diff
@@ -90,15 +117,35 @@ struct MantissaRange
...
                 return 1'000'000'000'000'000ULL;
+            case MantissaScale::LargeLegacy:
             case MantissaScale::Large:
                 return 1'000'000'000'000'000'000ULL;
             default:
-                // Since this can never be called outside a non-constexpr
-                // context, this throw assures that the build fails if an
+                // If called in a constexpr context, this throw assures that the build fails if an
+                // invalid scale is used.
+                throw std::runtime_error("Unknown mantissa scale");
```
Comment (Tapanito) [inline include/xrpl/basics/Number.h]: `codecov` is complaining that the `default` case is not tested. It seems to me testing this path would be non-trivial. Is it worth adding `//LCOV_EXCL_LINE?


```suggestion
                throw std::runtime_error("Unknown mantissa scale"); //LCOV_EXCL_LINE
```

File: include/xrpl/basics/Number.h [hunk 7]
```diff
@@ -203,7 +250,7 @@ concept Integral64 = std::is_same_v<T, std::int64_t> || std::is_same_v<T, std::u
  * amendments are enabled to determine which result to expect.
  *
  */
-class Number
+class Number final
```
Comment (Tapanito) [inline include/xrpl/basics/Number.h:253]: This change reminds me, someone threw this idea, to introduce `UNumber : Number` class to represent unsigned values. What do you think about that?

File: src/libxrpl/protocol/Rules.cpp [hunk 8]
```diff
@@ -38,15 +39,76 @@ setCurrentTransactionRules(std::optional<Rules> r)
...
+    // The improved accuracy that will come from enabling large
+    // mantissas is a goal separate from SAV and LP. It was originally
+    // only tied to those two amendments to avoid needing a new
+    // amendment of its own, and because they require that behavior.
+    // Because fixCleanup3_2_0 fixes a separate bug related to the large
+    // mantissas, that can take precedence and activate the large
+    // mantissas even in the absence of the other two amendments.
+    //
+    // If any new conditions with new amendments are added, those amendments must also be added to
+    // createGuards.
```
Comment (Tapanito) [inline src/libxrpl/protocol/Rules.cpp]: To be a devil's advocate, `Cleanup3_2_0` is a fix amendment, in technicality it should not introduce new features, right? 

Especially, since we're thinking of releasing it as `VoteBehaviour::DefaultYes`, releasing bigger ranges under a fix amendment doesn't seem correct. What do you think @ximinez ?

File: src/libxrpl/basics/Number.cpp [hunk 9]
```diff
@@ -51,10 +122,37 @@ Number::getMantissaScale()
...
+    // q /= 8 approximately == u / 10
+    q >>= 3;
+    // r = u - q * 10  approximately == u % 10
+    auto r = static_cast<unsigned>(u - ((q << 3) + (q << 1)));
+    // correction c is 1 if r >= 10 else 0
+    auto c = (r + 6) >> 4;
+    u = q + c;
+    r -= c * 10;
+    return r;
 }
```
Comment (Tapanito) [inline src/libxrpl/basics/Number.cpp:156]: Could you add a unit-test demonstrating the equivalence please?

File: src/libxrpl/basics/Number.cpp [hunk 10]
```diff
@@ -242,18 +374,58 @@ Number::Guard::doRoundUp(
...
+                // be impossible to recurse more than once, because once the mantissa is divided by
+                // 10, it will be _well_ under maxMantissa and kMAX_REP, so adding 1 will have no
+                // change of bringing it back over.
+                doDropDigit(mantissa, exponent);
+                XRPL_ASSERT_PARTS(
+                    mantissa < maxMantissa && mantissa < kMAX_REP,
+                    "xrpl::Number::Guard::doRoundUp",
+                    "can't recurse more than once");
+                // Here be dragons
+                doRoundUp(
```
Comment (Tapanito) [inline src/libxrpl/basics/Number.cpp:407]: What do you think about adding a depth guard to `doRoundUp` to ensure we really never recurse more than expected?

File: src/libxrpl/basics/Number.cpp [hunk 11]
```diff
@@ -365,7 +519,8 @@ doNormalize(
     T& mantissa,
     int& exponent,
     MantissaRange::rep const& minMantissa,
-    MantissaRange::rep const& maxMantissa)
+    MantissaRange::rep const& maxMantissa,
+    MantissaRange::CuspRoundingFix cuspRoundingFixEnabled)
```
Comment (Tapanito) [inline src/libxrpl/basics/Number.cpp:525]: Thought:

`min|maxMantissa` and `fixEnabled` all come from `MantissaRange` anyway, right? If so, could we simply refactor these methods to take `MantissaRange const&` instead?

**Review by Tapanito — CHANGES_REQUESTED** [review]
Every AMM test file subtracts the amendment from the feature set:

  - AMMTest.h — default TestAMMArg::features and AMMTestBase::features() both subtract fixCleanup3_2_0
  - AMMTest.cpp — testAMM() helper subtracts it again inside
  - AMMExtended_test.cpp — all_ member field subtracts it
  - AMMClawback_test.cpp — all local subtracts it
  - AMMClawbackMPT_test.cpp — the one combination that previously didn't now does too

How come? Especially since we're thinking of auto-enablng this with the cleanup amendment, we really need to ensure that it won't break the AMM.

File: include/xrpl/basics/Number.h (hunk 7 above)
Comment (ximinez) [reply include/xrpl/basics/Number.h:253]: > This change reminds me, someone threw this idea, to introduce `UNumber : Number` class to represent unsigned values. What do you think about that?

I've thought about something like that, but that's an enhancement to consider in the future. I can always remove `final` then if necessary.

File: src/libxrpl/protocol/Rules.cpp (hunk 8 above)
Comment (ximinez) [reply src/libxrpl/protocol/Rules.cpp]: > To be a devil's advocate, `Cleanup3_2_0` is a fix amendment, in technicality it should not introduce new features, right?
> 
> Especially, since we're thinking of releasing it as `VoteBehaviour::DefaultYes`, releasing bigger ranges under a fix amendment doesn't seem correct. What do you think @ximinez ?

Ugh. You're right. I didn't think of that. I'll change the priority.

Incidentally, that'll resolve your concern about the AMM tests. 

> How come? Especially since we're thinking of auto-enablng this with the cleanup amendment, we really need to ensure that it won't break the AMM.

The short answer is that enabling large mantissas in AMM tests breaks _hundreds_ of tests. Not because the computation is wrong, but because the computation results in different rounding, or more precision in some of the answers. The problem is that those tests use hard coded values as their expected results, so updating them will require conditionally updating all every single one of those values. It's not worth the time it would take to do that manually. Maybe I could use AI to do it more automatically now.

File: src/libxrpl/basics/Number.cpp (hunk 9 above)
Comment (ximinez) [reply src/libxrpl/basics/Number.cpp:156]: > Could you add a unit-test demonstrating the equivalence please?

This is not new code. I merely moved it from another location in this file. It's already pretty thoroughly tested by the `Number` tests, specifically multiplication. Notice that only one existing `Number` test had to change as a result of this fix, and that's because its rounding was fixed by this change.

Also, this function is not exposed, so writing direct tests is non-trivial.

File: src/libxrpl/basics/Number.cpp (hunk 10 above)
Comment (ximinez) [reply src/libxrpl/basics/Number.cpp:407]: > What do you think about adding a depth guard to `doRoundUp` to ensure we really never recurse more than expected?

Eh. I'm reluctant, because it really should be impossible, and checking would add overhead in potentially hot path.

I'll walk through it. Let's start with the assertion:
```
                XRPL_ASSERT_PARTS(
                    mantissa < maxMantissa && mantissa < kMaxRep,
                    "xrpl::Number::Guard::doRoundUp",
                    "can't recurse more than once");
```
Then `doRoundUp` is called with `mantissa`, `maxMantissa`, and `cuspRoundingFieldEnabled` amongst others.

So we enter `doRoundUp`, it makes the same `round()` decision (`r`), and enters the `if (cuspRoundingFixEnabled == MantissaRange::CuspRoundingFix::Enabled)` block. That starts with
```
            // Ensure mantissa after incrementing fits within both the
            // min/maxMantissa range and is a valid "rep".
            if (mantissa < maxMantissa && mantissa < kMaxRep)
            {
                // Nothing unusual here, just increment the mantissa
                ++mantissa;
            }
```
Look closely. That `if` condition is exactly the same as the check in `XRPL_ASSERT_PARTS`. Thus it's impossible to recurse again.

~I'll make this more obvious by setting that condition to a local variable.~ No wait, I can't do that, because `mantissa` changes, so the result of that value would change.

Given all this, do you think the assert is sufficient, or should I check and throw?

File: src/libxrpl/basics/Number.cpp (hunk 11 above)
Comment (ximinez) [reply src/libxrpl/basics/Number.cpp:525]: > Thought:
> 
> `min|maxMantissa` and `fixEnabled` all come from `MantissaRange` anyway, right? If so, could we simply refactor these methods to take `MantissaRange const&` instead?

Unfortunately, no. `normalizeToRange` also calls this function, and it can pass in arbitrary types and values. And a `MantissaRange` can only be constructed by one of the enum values, so I can't just make one on the fly.

Maybe I could make it a new struct that has more flexibility...

Comment (Tapanito) [reply src/libxrpl/basics/Number.cpp:525]: Aha, that's a shame. Let's not blow up the scope of the PR

File: src/libxrpl/basics/Number.cpp (hunk 10 above)
Comment (Tapanito) [reply src/libxrpl/basics/Number.cpp:407]: I don't think it's necessary to check and throw, thanks for walking me through that!

**Review by Tapanito — APPROVED** [review]
LGTM 🎉

File: src/test/app/AMMExtended_test.cpp [hunk 12]
```diff
@@ -65,10 +65,15 @@ namespace xrpl::test {
...
+class AMMExtended_test : public jtx::AMMTest
 {
     // Use small Number mantissas for the life of this test.
-    NumberMantissaScaleGuard const sg{xrpl::MantissaRange::MantissaScale::Small};
+    NumberMantissaScaleGuard const sg_{xrpl::MantissaRange::MantissaScale::Small};
+
+    // For now, just disable SAV entirely, which locks in the small Number
+    // mantissas
+    FeatureBitset const all_{
+        testableAmendments() - featureSingleAssetVault - featureLendingProtocol};
```
[the remark this answers is not in the document]
Comment (ximinez) [reply src/test/app/AMMExtended_test.cpp:76]: Good observation, wrong solution. I wanted to remove the `fixCleanup3_2_0` exclusion from all AMM tests, and I missed some. The solution is to remove them from the rest of those tests.

**Review by vvysokikh1 — APPROVED** [review]
lgtm!

File: src/libxrpl/basics/Number.cpp [hunk 13]
```diff
@@ -721,34 +856,53 @@ Number::operator/=(Number const& y)
     auto const& range = kRange.get();
     auto const& minMantissa = range.min;
     auto const& maxMantissa = range.max;
+    auto const cuspRoundingFixEnabled = range.cuspRoundingFixEnabled;
+
+    // Cache power of 10 computations to not waste time on subsequent calls
+    auto getPower10 = [](int exponent) {
+        static std::unordered_map<int, uint128_t> powersOf10;
```
[the remark this answers is not in the document]
Comment (vvysokikh1) [reply src/libxrpl/basics/Number.cpp]: forget about data race or something.

Why does this cache need to be static unordered map in the first place? There is at best 38 exponents that can represent values in uint128. Just make a constexpr array of all those values.

[the remark this answers is not in the document]
Comment (Tapanito) [reply src/libxrpl/basics/Number.cpp]: In the current call-site, there are only three inputs `{17, 19, 3}` :

```
 auto const factorExponent = small ? 17 : 19;
 uint128_t const f = getPower10(factorExponent);

...

bool const useBigCorrection =
            cuspRoundingFixEnabled == MantissaRange::CuspRoundingFix::Enabled;
auto const correctionExponent = useBigCorrection ? 17 : 3;         
```

While this is a very hot-path, a map seems like an overkill. Also, given the three possible inputs, they all seem to fit into `uint64`:

```
  ┌──────────┬────────────────────────────┬─────────────────────────────────────────────┐
  │ Exponent │           Value            │              Fits in uint64_t?              │
  ├──────────┼────────────────────────────┼─────────────────────────────────────────────┤
  │ 3        │ 1'000                      │ ✓                                           │
  ├──────────┼────────────────────────────┼─────────────────────────────────────────────┤
  │ 17       │ 100'000'000'000'000'000    │ ✓ (max uint64 ≈ 1.84×10^19)                 │
  ├──────────┼────────────────────────────┼─────────────────────────────────────────────┤
  │ 19       │ 10'000'000'000'000'000'000 │ ✓ (just barely — 10^19 < 2^64 ≈ 1.84×10^19) │
  └──────────┴────────────────────────────┴─────────────────────────────────────────────┘
```

What about having named constants at the top of operator/=:
```
  static constexpr uint128_t kPow10_3  = 1'000ULL;
  static constexpr uint128_t kPow10_17 = 100'000'000'000'000'000ULL;
  static constexpr uint128_t kPow10_19 = 10'000'000'000'000'000'000ULL;
```
  Then the two use sites read naturally:
```
  uint128_t const f = small ? kPow10_17 : kPow10_19;
  ...
  uint128_t const correctionFactor = useBigCorrection ? kPow10_17 : kPow10_3;
```

**Review by Tapanito — CHANGES_REQUESTED** [review]

File: src/libxrpl/basics/Number.cpp [hunk 14]
```diff
@@ -858,31 +858,51 @@ Number::operator/=(Number const& y)
...
+    // uint128_t or the cast back to int64_t for small mantissas.
+    //
+    // For large mantissas:
+    // * log(2^128,10) ~ 38.5
+    // * largeRange.log = 18, fits in 10^19
+    // * The expanded numerator must fit in 10^38
+    // * f can be up to 10^(38-19) = 10^19 safely
     bool const small = Number::getMantissaScale() == MantissaRange::MantissaScale::Small;
-    uint128_t const f = small ? 100'000'000'000'000'000 : 10'000'000'000'000'000'000ULL;
+    auto const factorExponent = small ? 17 : 19;
```
Comment (Tapanito) [inline src/libxrpl/basics/Number.cpp]: I suggest moving `17` and `19` to named variables rather than magic constants.

File: src/libxrpl/basics/Number.cpp [hunk 15]
```diff
@@ -858,31 +858,51 @@ Number::operator/=(Number const& y)
...
     bool zn = (ns * ds) < 0;
     if (!small)
     {
+        // zm may now hold up to 20 digits.  The correctionFactor must be small enough to not cause
+        // overflow, but as large as possible otherwise
+        // * zm fits in 10^21
+        // * correctionFactor can be up to 10^(38-21) = 10^17 safely
+        bool const useBigCorrection =
+            cuspRoundingFixEnabled == MantissaRange::CuspRoundingFix::Enabled;
+        auto const correctionExponent = useBigCorrection ? 17 : 3;
```
Comment (Tapanito) [inline src/libxrpl/basics/Number.cpp]: Ditto here, named variables would make the code easier to understand.

File: src/libxrpl/basics/Number.cpp
Comment (vvysokikh1) [inline src/libxrpl/basics/Number.cpp:516]: nit, can be ignored: this is not a part of this PR, but easy enough to cleanup this todo

File: include/xrpl/basics/Number.h [hunk 16]
```diff
@@ -424,49 +471,29 @@ class Number
...
-    // And one is needed because it needs to choose between oneSmall and
-    // oneLarge based on the current range
     static Number
     one();
 
-    template <Integral64 T>
+    template <
+        auto MinMantissa,
+        auto MaxMantissa,
+        Integral64 T = std::decay_t<decltype(MinMantissa)>,
```
Comment (vvysokikh1) [inline include/xrpl/basics/Number.h:480]: should this be TMin?

**Review by vvysokikh1 — COMMENTED** [review]
the fix itself seems legit, it just feels a bit rushed and brittle

**Comment by ximinez** [comment]
@Tapanito @vvysokikh1 Thanks for the reviews.

I realized over the long weekend that the last commit (48b1716e6f Make Number::operator/= significantly more accurate) should have been a separate, stacked PR. Even though it also has to do with rounding, it is a completely separate issue, and potentially more of an enhancement than a bug fix.

* #7328

That means that everything between https://github.com/XRPLF/rippled/pull/7051#pullrequestreview-4349035494 and this comment can more-or-less be ignored. (I'll copy anything salient over to the new PR.)

Where that leaves this PR:
* Approved by @vvysokikh1: https://github.com/XRPLF/rippled/pull/7051#pullrequestreview-4347183283
* A slightly outdated approval from @Tapanito: https://github.com/XRPLF/rippled/pull/7051#pullrequestreview-4335689590
* Some new Copilot feedback (dangit): https://github.com/XRPLF/rippled/pull/7051#pullrequestreview-4347707213
* Anything that might come up from CI.

**Review by Tapanito — APPROVED** [review]
With the reverted `=/` change: LGTM 🥳

File: include/xrpl/protocol/Rules.h [hunk 17]
```diff
@@ -122,4 +122,16 @@ class CurrentTransactionRulesGuard
...
+
+bool
+useRulesGuards(Rules const& rules);
+
+void
+createGuards(
+    Rules const& rules,
+    std::optional<NumberSO>& stNumberSO,
+    std::optional<CurrentTransactionRulesGuard>& rulesGuard,
+    std::optional<NumberMantissaScaleGuard>& mantissaScaleGuard);
```
[the remark this answers is not in the document]
Comment (ximinez) [reply include/xrpl/protocol/Rules.h:136]: Builds just fine for me without unity or anything else that would side-step the header requirements. Probably because the `std::optional` is a reference, so the type does not yet have to be complete.

To be safer, I'll predeclare `NumberMantissaScaleGuard`.

File: src/test/basics/Number_test.cpp [hunk 18]
```diff
@@ -19,6 +21,24 @@ namespace xrpl {
...
+        std::string out;
+        int count = 0;
+        for (auto it = s.rbegin(); it != s.rend(); ++it)
+        {
+            if (count != 0 && count % 3 == 0)
+                out.insert(out.begin(), '_');
+            out.insert(out.begin(), *it);
+            ++count;
+        }
+        return out;
```
[the remark this answers is not in the document]
Comment (ximinez) [reply src/test/basics/Number_test.cpp:40]: This is test code, which is more concerned with actual usage instead of potential future usage, and doesn't necessarily have to handle every edge case.

However, this looks like a pretty simple thing to fix, so I'll add an `isdigit` check.

File: src/test/basics/Number_test.cpp [hunk 19]
```diff
@@ -1551,11 +1585,49 @@ class Number_test : public beast::unit_test::Suite
...
+    {
+        testcase << "upward rounding produces a value below exact at kMaxRep cusp";
+
+        NumberMantissaScaleGuard const mg{MantissaRange::MantissaScale::Large};
+        NumberRoundModeGuard const rg{Number::RoundingMode::Upward};
+
+        constexpr std::int64_t kAValue = 1'000'000'000'000'049'863LL;
+        constexpr std::int64_t kBValue = 9'223'372'036'854'315'903LL;
+
+        // Public conversion operator: STAmount::operator Number() const.
```
[the remark this answers is not in the document]
Comment (ximinez) [reply src/test/basics/Number_test.cpp]: The comment doesn't add any value. Removed.

File: include/xrpl/basics/Number.h [hunk 20]
```diff
@@ -59,29 +61,54 @@ isPowerOfTen(T value)
...
     using rep = std::uint64_t;
-    enum class MantissaScale { Small, Large };
+    enum class MantissaScale {
+        Small,
+        // LargeLegacy can be removed when fixCleanup3_2_0 is retired
+        LargeLegacy,
+        Large,
+    };
+
+    // This entire enum can be removed when fixCleanup3_2_0 is retired
```
Comment (TimothyBanks) [inline include/xrpl/basics/Number.h:86]: not for this pr, but I was wondering if it might be useful to have something more targeted, like a `RETIRE_WITH` macro, that would make these sort of transient changes easier to search for.  The macro itself could just be a no-op or inject a `[[deprecated]]` annotation (although that would probably be noisy),

Comment (ximinez) [reply include/xrpl/basics/Number.h:86]: > not for this pr, but I was wondering if it might be useful to have something more targeted, like a `RETIRE_WITH` macro, that would make these sort of transient changes easier to search for. The macro itself could just be a no-op or inject a `[[deprecated]]` annotation (although that would probably be noisy),

Good question.

The intention of retiring features is to _remove_ code, so it can be simpler, easier to read, etc.

The process, amongst other things, removes the declaration from `Feature.h`, which causes any reference to it to fail to build. Including it in a comment won't have that effect, but it can at least be easier to find with a text search.

A macro that keeps the code seems to go in the wrong direction, unless maybe it's a macro that causes a compilation error when the feature is retired.

**Review by vvysokikh1 — APPROVED** [review]

File: include/xrpl/basics/Number.h (hunk 16 above)
Comment (bthomee) [reply include/xrpl/basics/Number.h:480]: @ximinez I was going to merge this PR but let me know if this is worth changing.

File: src/libxrpl/basics/Number.cpp [hunk 21]
```diff
@@ -242,18 +374,60 @@ Number::Guard::doRoundUp(
...
+                // Nothing unusual here, just increment the mantissa
+                ++mantissa;
+            }
+            else
+            {
+                // Incrementing the mantissa will require dividing, which will require rounding. So
+                // _don't_ increment the mantissa. Instead, divide and round recursively. It should
+                // be impossible to recurse more than once, because once the mantissa is divided by
+                // 10, it will be _well_ under maxMantissa and kMaxRep, so adding 1 will have no
+                // change of bringing it back over.
```
Comment (bthomee) [inline src/libxrpl/basics/Number.cpp:401]:
```suggestion
                // chance of bringing it back over.
```

This word change is probably not worth making, unless you're going to also change `T` to `TMin` as suggested by Vlad.

File: include/xrpl/basics/Number.h (hunk 16 above)
Comment (ximinez) [reply include/xrpl/basics/Number.h:480]: > I was going to merge this PR but let me know if this is worth changing.

@bthomee I'll handle it over on #7328. My initial feedback there was "Response: Both values are supposed to have the same type: T. TMax is only used to verify that. I can add a comment and/or rename TMax to TConfirm."

File: src/libxrpl/basics/Number.cpp
Comment (ximinez) [reply src/libxrpl/basics/Number.cpp:516]: > nit, can be ignored: this is not a part of this PR, but easy enough to cleanup this todo

Done in #7328.

File: src/libxrpl/basics/Number.cpp (hunk 21 above)
Comment (ximinez) [reply src/libxrpl/basics/Number.cpp:401]: Fixed in #7328.
